src/NuGet.Core/NuGet.Commands/CommandRunners/PackageSearchRunner.cs

```diff
@@ -0,0 +1,198 @@
+        {
+            var taskList = new List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)>();
+
+            foreach (var source in listEndpoints)
+            {
+                var repository = Repository.Factory.GetCoreV3(source);
+                var resource = await repository.GetResourceAsync<PackageSearchResource>(cancellationToken);
+                taskList.Add(
+                    resource == null ? (null, source) :
+                    (resource.SearchAsync(searchTerm, new SearchFilter(includePrerelease: prerelease), skip, take, logger, cancellationToken), source)
+                );
+            }
+
+            return taskList;
+        }
+
+        /// <summary>
+        /// Builds a list of tasks that perform exact match package search operations.
+        /// </summary>
+        /// <param name="listEndpoints">The list of package sources/endpoints.</param>
+        /// <param name="searchTerm">The term to search for within the package sources.</param>
+        /// <param name="prerelease">A flag indicating whether to include prerelease packages in the search.</param>
+        /// <param name="logger">The logger instance to use for logging.</param>
+        /// <param name="cancellationToken">A token to monitor for cancellation requests.</param>
+        /// <returns>A list of tasks that perform exact match package search operations.</returns>
+        private static async Task<List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)>> BuildTaskListForExactMatch(
+            IList<PackageSource> listEndpoints,
+            string searchTerm,
+            bool prerelease,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var taskList = new List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)>();
+
+            foreach (var endpoint in listEndpoints)
+            {
+                using var cache = new SourceCacheContext();
+                var repository = Repository.Factory.GetCoreV3(endpoint);
+                var resource = await repository.GetResourceAsync<PackageMetadataResource>();
+                taskList.Add(
+                    resource == null ? (null, endpoint) :
+                    (resource.GetMetadataAsync(searchTerm, includePrerelease: prerelease, includeUnlisted: false, cache, logger, cancellationToken), endpoint)
+                );
+            }
+
+            return taskList;
+        }
+
+        /// <summary>
+        /// Retrieves a list of package sources based on provided sources or default configuration sources.
+        /// </summary>
+        /// <param name="sources">The list of package sources provided.</param>
+        /// <param name="sourceProvider">The provider for package sources.</param>
+        /// <returns>A list of package sources.</returns>
+        private static IList<PackageSource> GetEndpointsAsync(List<string> sources, IPackageSourceProvider sourceProvider)
```

Comment:
```suggestion
        private static IList<PackageSource> GetPackageSourcesAsync(List<string> sources, IPackageSourceProvider sourceProvider)
```

I don't believe we use the term "endpoint" much (at all?) in NuGet, so calling it "package sources" reuses existing common vocabulary, and therefore is faster to read for people who are familar with NuGet's common vocabulary.